hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/update/client/CRLClientUpdateHandler.java

```diff
@@ -0,0 +1,201 @@
+  private SCMUpdateServiceProtos.ClientId clientIdProto;
+
+  // periodically process pending crls
+  private ScheduledExecutorService executorService;
+  private final SCMUpdateServiceGrpcClient serviceGrpcClient;
+  private long crlCheckInterval;
+
+  CRLClientUpdateHandler(UUID clientId,
+      SCMUpdateServiceStub updateStub,
+      SCMUpdateServiceGrpcClient serviceGrpcClient,
+      long crlCheckInterval) {
+    this.clientUuid = clientId;
+    this.updateStub = updateStub;
+    this.serviceGrpcClient = serviceGrpcClient;
+
+    this.clientStore = serviceGrpcClient.getClientCRLStore();
+    this.crlCheckInterval = crlCheckInterval;
+    LOG.info("Pending CRL check interval : {}s", crlCheckInterval/1000);
+    this.executorService = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setDaemon(true)
+            .setNameFormat("CRLUpdateHandler Thread - %d").build());
+  }
+
+  public static Logger getLog() {
+    return LOG;
+  }
+
+  @Override
+  public void handleServerUpdate(UpdateResponse updateResponse) {
+    SCMUpdateServiceProtos.CRLInfoProto crlIfo =
+        updateResponse.getCrlUpdateResponse().getCrlInfo();
+
+    long receivedCrlId = crlIfo.getCrlSequenceID();
+    long localCrlId = clientStore.getLatestCrlId();
+
+    LOG.debug("## Client: clientId {} clientCrlId {} receivedCrlId {}",
+        clientUuid, localCrlId, receivedCrlId);
+    if (localCrlId == receivedCrlId) {
+      return;
+    }
+    // send a client update to refresh stale server
+    if (localCrlId > receivedCrlId) {
+      LOG.warn("Receive stale crlId {} lower than client crlId {}",
+          receivedCrlId, localCrlId);
+      sendClientUpdate();
+      return;
+    }
+
+    CRLInfo crl;
+    try {
+      crl = CRLInfo.fromCRLProto3(crlIfo);
+    } catch (Exception e) {
+      LOG.error("Can't parse server CRL update, skip...", e);
+      return;
+    }
+    clientStore.onRevokeCerts(crl);
+    // send client update.
+    sendClientUpdate();
+  }
+
+  public void start() {
+    // send initial update request to get a request observer handle
+    UpdateRequest updateReq = getUpdateRequest();
+    requestObserver = updateStub.withWaitForReady()
+        .updateStatus(new StreamObserver<UpdateResponse>() {
+          @Override
+          public void onNext(UpdateResponse updateResponse) {
+            LOG.debug("Receive server response: {}", updateResponse.toString());
+            serviceGrpcClient.incrUpdateCount();
+            handleServerUpdate(updateResponse);
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            LOG.debug("Receive server error ", throwable);
+            serviceGrpcClient.incrErrorCount();
+            if (serviceGrpcClient.getIsRunning().get()) {
+              // TODO: not all server error needs client restart.
+              LOG.warn("Restart client on server error: ", throwable);
+              serviceGrpcClient.restart();
+            }
```

Comment:
The assumption here is to avoid endless loop on client restart, i.e., when the client is shutdown(not running), we will not try to restart client to handle server errors.